### ISSUE TITLE
Anderson-Darling ksamples can not estimate p-values beyond given range

### ISSUE THREAD

**Issue by Gabriel-p** [scipy/scipy#9527]
I've been following the trace for this long standing issue. let me do a quick recap:

It was apparently first reported in 2015 (https://github.com/scipy/scipy/issues/5493) mentioning that the test was throwing bad p-values (larger than 1). A similar issue was reported this year (https://github.com/scipy/scipy/issues/8536) with the discussion apparently agreeing on capping the p-values at 0.25. This was addressed by this commit https://github.com/scipy/scipy/pull/8560.

Both issues above are now closed, but I think the problem remains. Capping the p-values at 0.25 sort of works, but it does not "fix" the test. My main question is (this also came up in the threads mentioned above): why can't we do whatever is done by the [kSamples](https://cran.r-project.org/web/packages/kSamples/index.html) R package?

I did some tests with the `ad.test` function in that package, and it does not have the problems with the p-values that [anderson_ksamp](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.anderson_ksamp.html) has.

Can't whatever is done in `ad.test` be incorporated into `anderson_ksamp`?

**Comment by josef-pkt** [comment]
issue 8536 has some discussion about extending the range of p-values.

Someone has to implement the p-values or p-value tables, but we cannot translate or directly use the implementation in the kSamples package because it is GPL and not license compatible.

@aarchiba has an implementation in https://github.com/aarchiba/kuiper/blob/master/anderson_darling.py
but I never tried it out. 
(It has BSD-3 license, but didn't have it at the time I was looking at gof tests)

**Comment by Gabriel-p** [comment]
Thanks @josef-pkt. I did some quick tests with that code, but it does not seem to work entirely. This;

```
from scipy.stats import anderson_ksamp

a = [-1.8, -2.4, -2.4, -0.0, -1.5, -2.7, -1.8, -3.0, -1.8, -1.2, -3.0, -3.0, -2.8, -3.0, -2.1, -0.0, 0.6, -2.5, -2.4, -0.0, -2.7, -0.0, -2.5, -2.1, -0.9, -3.0, -0.6, -0.6, -1.5, -2.2, -1.2, -2.4, -2.4, -3.0, 1.5, -1.8, 1.5, -2.7, -3.0, -2.5, -2.5, -1.5, -1.5, -2.1, -2.1, -3.0, -0.6, -2.7, -3.0, -1.5, -0.6, -0.0, -2.1, 0.6, -2.0, -3.0, -3.0, -2.4, -3.0, -1.8, -0.0, -0.0, -3.0, -1.5, -3.0, -3.0, -1.5, -2.5, -3.0, -2.8, -3.0, -2.2, -0.6, -0.0, -1.5, -2.7, -2.1, -2.1, -2.2, -2.1, -0.6, -0.0, -2.5, -2.1, -1.5, -3.0, -2.2, -1.8, -2.7, -2.4, -1.5, -2.1, -2.9, -2.4, -0.9, -0.0, -0.0, -2.4, -2.7, -0.6, -2.2, -3.0, -1.5, -0.9, -3.0, -3.0, -0.0, -2.7, -2.7, -1.5, -2.2, -3.0, -0.0, 1.5, -3.0, -2.7, -2.2, -2.9, -2.2, -3.0, -1.8, -0.0, -3.0, -1.5, -2.7, -3.0, -3.0, -2.9, -3.0, -3.0, -3.0, -3.0, -2.5, -0.0, 0.9, -3.0, -0.0, -3.0, 3.0, -3.0, -3.0, -1.2, -2.1, 1.5, -0.0, -0.9, -3.0, -2.7, -1.5, -2.4, -2.1, -3.0, -0.9, -3.0, -0.8, -1.5, -2.1, -2.7, -0.0, -0.0, -2.2, -1.8, -2.1, -2.2, -3.0, 0.6, -2.4, -2.2, -2.4, -2.5, -1.5, -0.0, -2.7, -3.0, -3.0, -2.1, -0.0, -2.4, -2.4, -0.0, -2.0, -0.9, -2.4, -3.0, -1.4, -2.7, -2.7, -3.0, -3.0, -2.7, -1.2, -2.1, -3.0, -0.0, -3.0, -2.7, -2.7, -3.0, -3.0, -2.5, -3.0, -1.8, -1.5, -2.7, -2.4, -1.8, -3.0, -2.7, 2.1, -3.0, -2.2, -2.2, 0.6, -0.9, 6.0, -3.0, -2.1, -3.0, -2.1, -2.5, -3.0, -1.5, -2.5, 3.0, -2.1, -3.0, -3.0, -1.5, -2.1, -2.7, -2.5, 1.5, -2.1, -0.0, -0.0, -3.0, -0.6, 1.5, -2.7, -2.4, -2.1, -3.0, -2.7, -3.0, 9.0, -3.0, -1.7, -3.0, -0.0, -3.0, -2.2, -0.0, -0.6, -2.7, -2.7, -3.0, -3.0, -1.7, -2.1, -2.0, -3.0, -2.1, -3.0, -1.1, -3.0, -3.0, -2.4, -1.5, -3.0, -2.2, -3.0, -1.5, -2.7, -3.0, -3.0, -3.0, -2.2, -3.0, -2.1, -2.1, -2.4, -3.0, -3.0, -0.0, -3.0, -3.0, -2.1, -2.0, 1.5, -3.0, -3.0, -2.1, -2.9, -2.4, -3.0, -3.0, -1.5, -2.2, -0.9, -1.8, -2.1, -1.8, -1.5, -3.0, -1.5, -3.0, -1.5, -3.0, -2.4, 1.5, -2.7, -3.0, -1.8, -1.8, -1.5, -2.1, -2.7, -2.7, -2.7, -3.0, -1.5, -2.7, -3.0, -2.7, -3.0, -1.5, -1.5, -3.0, 0.6, -0.6, -3.0, -2.1, -2.4, -2.1, -3.0, -2.2, -3.0, -1.8, -1.2, -3.0, -0.8, -2.4, -2.5, -3.0, -1.5, -1.2, -0.0, -2.7, -2.4, -3.0, -3.0, -2.1, -2.1, -2.1, -3.0, -2.7, -2.4, -2.1, -1.5, -2.1, -0.6, -3.0, -3.0, -3.0, -3.0]
b = [-1.8, -2.4, -2.4, -0.0, -1.5, -2.7, -1.8, -3.0, -1.8, -1.2]

anderson_ksamp([a, b])
```

correctly returns:

> Anderson_ksampResult(statistic=0.12824285083144032, critical_values=array([0.325, 1.226, 1.961, 2.718, 3.752]), significance_level=0.30477949380558533)

but the `anderson_darling_k()` function returns:

> (1.1023824662400878, 0.1339627318973063, (array([0.325, 1.226, 1.961, 2.718, 3.752]), array([0.25 , 0.1  , 0.05 , 0.025, 0.01 ])), 1.0)

These values coincide with the result from the `ad.test()` function (`version 1`  results) from `kSamples`, except for the p-value:

>  Anderson-Darling k-sample test.
> 
> Number of samples:  2
> Sample sizes:  365, 10
> Number of ties: 349
> 
> Mean of  Anderson-Darling  Criterion: 1
> Standard deviation of  Anderson-Darling  Criterion: 0.76426
> 
> T.AD = ( Anderson-Darling  Criterion - mean)/sigma
> 
> Null Hypothesis: All samples come from a common population.
> 
>               AD   T.AD  asympt. P-value
> version 1: 1.102 0.1340           0.3076
> version 2: 1.100 0.1282           0.3096

**Comment by Gabriel-p** [comment]
Thanks @chrisb83. So I can use that gist to extend the coefficients to obtain the p-value for 0.5, 0.75, and 1 for example? This seems like an easy enough task, is there any specific reason it isn't done already? I mean something like stability issues in the coefficients or too much time wasted on simulations.

**Comment by chrisb83** [comment]
Yes, I could have easily included other significance levels at that time, I did not consider larger significance levels though. For 0.5 and 0.75, the simulation is much easier since you are not far in the tail as for p=0.001, so probably one could use much less simulations. then you still need to fit the interpolation for the number of samples. Runtime of the script was ~12h (I used a very simple approach).

So actually, the extra effort is not too much if I think about it again :)

**Comment by mdhaber** [comment]
@Gabriel-p @chrisb83 Just thought I'd post how we can do this with `monte_carlo_test`. Not sure if we'd want to use it to compute p-values on the fly or precompute tables.

```python3
import numpy as np
from scipy import stats

rng = np.random.default_rng(3)

size = 10000
sample = rng.normal(size=size)

def statistic(sample):
    return stats.anderson(sample, dist='norm')[0]

res = stats.monte_carlo_test(sample, rng.normal, statistic, n_resamples=9999)
alpha = np.array([0.5, 0.25, 0.15, 0.1, 0.05, 0.025, 0.01])  # 
percentile = (1 - alpha)*100
critical_values = stats.scoreatpercentile(res.null_distribution, percentile)
print(critical_values)
```
prints
```
[0.3392179  0.46656921 0.56401377 0.63282984 0.73645927 0.85842631
 1.03623988 1.18666595]
```
Compare against those in [D'Agostino](https://www.lexjansen.com/nesug/nesug91/NESUG91049.pdf):

![image](https://user-images.githubusercontent.com/6570539/158522727-3eb9d5f2-ca61-4014-b604-e91c43c2ab9a.png)

And for exponential data:
```python3
sample = rng.exponential(size=size)

def statistic(sample):
    return stats.anderson(sample, dist='expon')[0]

res = stats.monte_carlo_test(sample, rng.exponential, statistic, n_resamples=9999)
alpha = np.array([0.25, 0.15, 0.1, 0.05, 0.025, 0.01])
percentile = (1 - alpha)*100
critical_values = stats.scoreatpercentile(res.null_distribution, percentile)
print(critical_values)
```
we get:
```
[0.73078297 0.90329669 1.04901069 1.30703602 1.56352514 1.93205574]
```
vs last row of
![image](https://user-images.githubusercontent.com/6570539/158525813-20ceacb1-8016-42d1-9c55-e7489be9cce9.png)

**Comment by mdhaber** [comment]
Oops, my post above was about `anderson`, but this issue is about `anderson_ksamp`.

TLDR: I could open a very simple PR to make `anderson_ksamp` calculate accurate p-values on the fly using `permutation_test`. It would add two parameters: `n_resamples=None` and `random_state=None`. 

1. By default, `anderson_ksamp`'s behavior would not change at all. In a follow-up PR, we could consider changing something about the default behavior (e.g. use a permutation test if the statistic is beyond the range of tabulated values.)
2. If `n_resamples` is not `None`, the p-value would be computed using `permutation_test` instead of interpolating. (We could have a default value for `n_resamples`, but then there would be ambiguity about what the user wants when `random_state` is `None`.)
3. `random_state` would simply be passed through to `permutation_test`.

@chrisb83 Let me know if you'd be willing to review, and I'd be happy to open a PR. It would be pretty simple since it won't involve any algorithmic code.

<details>

`anderson_ksamp` p-values can be calculated using a permutation test (exact or randomized).

For the example above:
```python3
from scipy import stats

x = [-1.8, -2.4, -2.4, -0.0, -1.5, -2.7, -1.8, -3.0, -1.8, -1.2, -3.0, -3.0, -2.8, -3.0, -2.1, -0.0, 0.6, -2.5, -2.4, -0.0, -2.7, -0.0, -2.5, -2.1, -0.9, -3.0, -0.6, -0.6, -1.5, -2.2, -1.2, -2.4, -2.4, -3.0, 1.5, -1.8, 1.5, -2.7, -3.0, -2.5, -2.5, -1.5, -1.5, -2.1, -2.1, -3.0, -0.6, -2.7, -3.0, -1.5, -0.6, -0.0, -2.1, 0.6, -2.0, -3.0, -3.0, -2.4, -3.0, -1.8, -0.0, -0.0, -3.0, -1.5, -3.0, -3.0, -1.5, -2.5, -3.0, -2.8, -3.0, -2.2, -0.6, -0.0, -1.5, -2.7, -2.1, -2.1, -2.2, -2.1, -0.6, -0.0, -2.5, -2.1, -1.5, -3.0, -2.2, -1.8, -2.7, -2.4, -1.5, -2.1, -2.9, -2.4, -0.9, -0.0, -0.0, -2.4, -2.7, -0.6, -2.2, -3.0, -1.5, -0.9, -3.0, -3.0, -0.0, -2.7, -2.7, -1.5, -2.2, -3.0, -0.0, 1.5, -3.0, -2.7, -2.2, -2.9, -2.2, -3.0, -1.8, -0.0, -3.0, -1.5, -2.7, -3.0, -3.0, -2.9, -3.0, -3.0, -3.0, -3.0, -2.5, -0.0, 0.9, -3.0, -0.0, -3.0, 3.0, -3.0, -3.0, -1.2, -2.1, 1.5, -0.0, -0.9, -3.0, -2.7, -1.5, -2.4, -2.1, -3.0, -0.9, -3.0, -0.8, -1.5, -2.1, -2.7, -0.0, -0.0, -2.2, -1.8, -2.1, -2.2, -3.0, 0.6, -2.4, -2.2, -2.4, -2.5, -1.5, -0.0, -2.7, -3.0, -3.0, -2.1, -0.0, -2.4, -2.4, -0.0, -2.0, -0.9, -2.4, -3.0, -1.4, -2.7, -2.7, -3.0, -3.0, -2.7, -1.2, -2.1, -3.0, -0.0, -3.0, -2.7, -2.7, -3.0, -3.0, -2.5, -3.0, -1.8, -1.5, -2.7, -2.4, -1.8, -3.0, -2.7, 2.1, -3.0, -2.2, -2.2, 0.6, -0.9, 6.0, -3.0, -2.1, -3.0, -2.1, -2.5, -3.0, -1.5, -2.5, 3.0, -2.1, -3.0, -3.0, -1.5, -2.1, -2.7, -2.5, 1.5, -2.1, -0.0, -0.0, -3.0, -0.6, 1.5, -2.7, -2.4, -2.1, -3.0, -2.7, -3.0, 9.0, -3.0, -1.7, -3.0, -0.0, -3.0, -2.2, -0.0, -0.6, -2.7, -2.7, -3.0, -3.0, -1.7, -2.1, -2.0, -3.0, -2.1, -3.0, -1.1, -3.0, -3.0, -2.4, -1.5, -3.0, -2.2, -3.0, -1.5, -2.7, -3.0, -3.0, -3.0, -2.2, -3.0, -2.1, -2.1, -2.4, -3.0, -3.0, -0.0, -3.0, -3.0, -2.1, -2.0, 1.5, -3.0, -3.0, -2.1, -2.9, -2.4, -3.0, -3.0, -1.5, -2.2, -0.9, -1.8, -2.1, -1.8, -1.5, -3.0, -1.5, -3.0, -1.5, -3.0, -2.4, 1.5, -2.7, -3.0, -1.8, -1.8, -1.5, -2.1, -2.7, -2.7, -2.7, -3.0, -1.5, -2.7, -3.0, -2.7, -3.0, -1.5, -1.5, -3.0, 0.6, -0.6, -3.0, -2.1, -2.4, -2.1, -3.0, -2.2, -3.0, -1.8, -1.2, -3.0, -0.8, -2.4, -2.5, -3.0, -1.5, -1.2, -0.0, -2.7, -2.4, -3.0, -3.0, -2.1, -2.1, -2.1, -3.0, -2.7, -2.4, -2.1, -1.5, -2.1, -0.6, -3.0, -3.0, -3.0, -3.0]
y = [-1.8, -2.4, -2.4, -0.0, -1.5, -2.7, -1.8, -3.0, -1.8, -1.2]

def statistic(*args):
    return stats.anderson_ksamp(args).statistic

res = stats.permutation_test((x, y), statistic, alternative='greater')
print(res.pvalue)  # 0.3161  # may vary
```

With smaller samples, we can do an exact test.
```python3
u1 = (1.0066, -0.9587,  0.3462)
u2 = (0.1005, 0.2252, 0.4810, 0.6992)
u3 = (-0.7019, -0.4083, -0.9936)
data = (u1, u2, u3)

def statistic(*data):
    return stats.anderson_ksamp(data, midrank=False).statistic

res2 = stats.permutation_test(data, statistic, alternative='greater', n_resamples=np.inf)
print(res2.pvalue)  # 0.16380952380952382
```

This matches the version 1 p-value in R's `library(kSamples)` `ad.test`.

```R
library(kSamples)
options(digits=16)
u1 = c(1.0066, -0.9587,  0.3462)
u2 = c(0.1005, 0.2252, 0.4810, 0.6992)
u3 = c(-0.7019, -0.4083, -0.9936)
ad.test(u1, u2, u3, method = "exact", dist = FALSE, Nsim=100000000.)

#               AD   T.AD  asympt. P-value  exact P-value
# version 1: 2.832 0.9867           0.1397         0.1638
# version 2: 2.860 1.0180           0.1352         0.1624
```

For most data I've tested, I can get exact agreement with R's `ad.test` when I force both functions to produce exact p-values, but they disagree in one case I've tried.
```python3
u1 = (1.0066, -0.9587,  0.3462, -0.2653)
u2 = (0.1005, 0.2252, 0.4810, 0.6992)
u3 = (-0.7019, -0.4083, -0.9936, -0.5439)
data = (u1, u2, u3)

res2 = stats.permutation_test(data, statistic, alternative='greater', n_resamples=np.inf)
print(res2.pvalue)  # 0.07411255411255412
```

In R:
```R
library(kSamples)
options(digits=16)
u1 = c(1.0066, -0.9587,  0.3462, -0.2653)
u2 = c(0.1005, 0.2252, 0.4810, 0.6992)
u3 = c(-0.7019, -0.4083, -0.9936, -0.5439)
ad.test(u1, u2, u3, method = "exact", dist = FALSE, Nsim=100000000.)
#              AD  T.AD  asympt. P-value  exact P-value
# version 1: 3.505 1.708          0.06446        0.07406
# version 2: 3.500 1.707          0.06458        0.07786
```


I'm inclined to trust `permutation_test` because it's been tested extensively, and there are some other reasons why the R version is suspect. 

0. `ad.test` does not respond consistently to `options(digits=16)`, which is weird.
1. `stats.anderson_ksamp` matches the version 2 T.AD statistic produced by `ad.test`, yet the permutation test agrees with the version 1 `pvalue` produced by `ad.test`. This suggests that the R version is mixing up the pvalues between the two versions.
5. Consider the test case below:

```R
library(kSamples)
options(digits=16)
u1 = c(1.0066, -0.9587,  0.3462, -0.2653, -1.3872)
u2 = c(0.1005, 0.2252, 0.4810, 0.6992, 1.9289)
u3 = c(-0.7019, -0.4083, -0.9936, -0.5439)
ad.test(u1, u2, u3, method = "exact", dist = FALSE, Nsim=100000000.)
```
The version 1 exact p-value is `0.05995000000000000`.
```
              AD  T.AD  asympt. P-value       exact P-value
version 1: 3.664 1.832          0.05633 0.05995000000000000
version 2: 3.650 1.814          0.05756 0.06635000000000001
```
The number of possible arrangements of ranks for $n_1=5$, $n_2=5$, $n_3=4$ is $14! / (5! 5! 4!) = 252252$, so this suggests that `ad.test`. observed $0.05995 \cdot 252252 = 15122.5074$ elements of the null distribution with a p-value equal to or exceeding 1.814. That doesn't make sense. The p-value multiplied by the number of elements in the null distribution must be an integer.

Likewise, in the discrepant example above, the size of the exact null distribution is 34650. $2566/34650=0.07405483405483405$ and $2567/34650=0.07408369408369408$ bound the p-value produced by `ad.test`, but neither of them round to the p-value reported by `ad.test`.

And just to check that `permutation_test` is doing the right thing, here is a stand-alone calculation of the exact null distribution in the case where we disagree with `ad.test`.
```python3
from scipy import stats
from itertools import combinations

u1 = (1.0066, -0.9587,  0.3462, -0.2653)
u2 = (0.1005, 0.2252, 0.4810, 0.6992)
u3 = (-0.7019, -0.4083, -0.9936, -0.5439)
observed_statistic = stats.anderson_ksamp((u1, u2, u3), midrank=False).statistic

# equivalently, we could form the null distribution with ranks
# u1 = (1, 2, 3, 4)
# u2 = (5, 6, 7, 8)
# u3 = (9, 10, 11, 12)

pool = set(u1 + u2 + u3)

n1 = len(u1)
n2 = len(u2)
n3 = len(u3)

null_dist = []
for x1 in combinations(pool, n1):
    remaining_pool = pool - set(x1)
    for x2 in combinations(remaining_pool, n2):
        x3 = tuple(remaining_pool - set(x2))
        data = (x1, x2, x3)
        stat = stats.anderson_ksamp(data, midrank=False).statistic
        null_dist.append(stat)
    
null_dist = np.array(null_dist)
N = len(null_dist)
pvalue_exact_arithmetic =np.sum(null_dist >= observed_statistic) / N
true_pvalue =np.sum(null_dist >= observed_statistic - 1e-15) / N
print(pvalue_exact_arithmetic, true_pvalue)  # 0.07399711399711399 0.07411255411255412
```
`true_pvalue` is what `permutation_test` produced. We need to subtract that small value from `observed_statistic` to perform the comparison because some values in the null distribution are numerically distinct even though they should be indistinguishable if  `anderson_ksamp` could compute the statistic with perfect accuracy.

</details>

**Comment by j-bowhay** [comment]
Reopening after #18321
